EIPS/EIP-X-metadata-nickjohnson.md

```diff
@@ -0,0 +1,72 @@
+
+## Specification
+The metadata registry has the following interface:
+```
+interface AddressMetadataRegistry {
+  function provider(address target) view returns(address);
+  function setProvider(address _provider);
+}
+```
+
+`setProvider` specifies the metadata registry to be associated with the caller's address, while `provider` returns the address of the metadata registry for the supplied address.
+
+The metadata registry will be compiled with an agreed-upon version of Solidity and deployed using the trustless deployment mechanism to a fixed address that can be replicated across all chains.
+
+## Provider specification
+
+Providers may implement any subset of the metadata record types specified here. Where a record types specification requires a provider to provide multiple functions, the provider MUST implement either all or none of them. Providers MUST throw if called with an unsupported function ID.
+
+Providers have one mandatory function:
+
+```
+function supportsInterface(bytes4 interfaceID) constant returns (bool)
+```
+
+The `supportsInterface` function is documented in [EIP 165](https://github.com/ethereum/EIPs/issues/165), and returns true if the provider implements the interface specified by the provided 4 byte identifier. An interface identifier consists of the XOR of the function signature hashes of the functions provided by that interface; in the degenerate case of single-function interfaces, it is simply equal to the signature hash of that function. If a provider returns `true` for `supportsInterface()`, it must implement the functions specified in that interface.
+
+`supportsInterface` must always return true for `0x01ffc9a7`, which is the interface ID of `supportsInterface` itself.
+
+The first argument to all provider functions MUST be the address being queried; this facilitates the creation of multi-user provider contracts.
+
+Currently standardised provider interfaces are specified in the table below.
+
+| Interface name | Interface hash | Specification |
+| --- | --- | --- |
+
+EIPs may define new interfaces to be added to this registry.
+
+## Rationale
+There are two obvious approaches for a generic metadata registry: the indirection approach employed here, or a generalised key/value store. While indirection incurs the cost of an additional contract call, and requires providers to change over time, it also provides for significantly enhanced flexibility over a key/value store; for that reason we selected this approach.
+
+## Backwards Compatibility
+There are no backwards compatibility concerns.
+
+## Implementation
+The canonical implementation of the metadata registry is as follows:
+```
+contract AddressMetadataRegistry {
+  mapping(address=>address) public provider;
+  
+  function setProvider(address _provider) {
+    provider[msg.sender] = _provider;
+  }
+}
+```
```

Comment:
Recommend including an example provider as well that at _least_ implements the required `supportsInfercae` code (and shows how to do the function signature hashing in Solidity).

Reply:
Agreed, though callers don't need to do the function hashing, just embed the appropriate constants (unless we can trust solidity to optimise those out, even if optimisation is turned off?)

Reply:
I think there is value in showing a user how to do the XOR, hashing and byte chopping in code, and Solidity is the language that we can be certain all readers of these understand.

Reply:
If Solidity doesn't optimise-away the hashing at compile-time, though, this is going to be a significant waste of gas.

Reply:
I would be OK (though not quite as satisfied) with an example written in JS if you think that is more reasonable.  Solidity is just the "common language".  I would not recommend people actually do the hashing in Solidity with each contract call.

Reply:
As a developer who didn't know of EIP 165's prior to reading this EIP, I was surprised to find it lacking an example for the "other side".  If you think this EIP should depend on 165, then I recommend including such a dependency in the EIP and indicating how the two are connected to each other.  In general, if an EIP doesn't reference any other EIPs, then I assume it can be consumed on its own (which is what I tried to do when reading this EIP) and it seems that your intent is that this EIP is _not_ consumed on its own.

Reply:
And of course after saying that, I re-read this EIP and found it did mention EIP 165...